packages/core/src/modules/drpc/DRPCMessagesApi.ts

```diff
@@ -0,0 +1,173 @@
+import type { DRPCRequest, DRPCRequestObject, DRPCResponse, DRPCResponseObject } from './messages'
+import type { DRPCMessageRecord } from './repository/DRPCMessageRecord'
+import type { ConnectionRecord } from '../connections'
+
+import { AgentContext } from '../../agent'
+import { MessageHandlerRegistry } from '../../agent/MessageHandlerRegistry'
+import { MessageSender } from '../../agent/MessageSender'
+import { OutboundMessageContext } from '../../agent/models'
+import { injectable } from '../../plugins'
+import { isValidDRPCRequestObject } from '../../utils/messageType'
+import { uuid } from '../../utils/uuid'
+import { ConnectionService } from '../connections'
+
+import { DRPCMessageRole } from './DRPCMessageRole'
+import { DRPCMessageHandler } from './handlers'
+import { DRPCErrorCode, DRPCRequestMessage, DRPCResponseMessage } from './messages'
+import { DRPCMessageService } from './services'
+
+@injectable()
+export class DRPCMessagesApi {
+  private drpcMessageService: DRPCMessageService
+  private drpcMethodHandlers: Map<
+    string,
+    (message: DRPCRequestObject) => Promise<DRPCResponseObject | Record<string, never>>
+  > = new Map()
+  private messageSender: MessageSender
+  private connectionService: ConnectionService
+  private agentContext: AgentContext
+
+  public constructor(
+    messageHandlerRegistry: MessageHandlerRegistry,
+    drpcMessageService: DRPCMessageService,
+    messageSender: MessageSender,
+    connectionService: ConnectionService,
+    agentContext: AgentContext
+  ) {
+    this.drpcMessageService = drpcMessageService
+    this.messageSender = messageSender
+    this.connectionService = connectionService
+    this.agentContext = agentContext
+    this.registerMessageHandlers(messageHandlerRegistry)
+
+    // listen for incoming drpc requests and forward them to appropriate handlers
+    this.drpcMessageService.createMessageListener(async ({ message, drpcMessageRecord }) => {
```

Comment:
For this particular case I don't see why it is needed to add a listener. It can be handled directly within `DrpcMessageService` in a `processRequest` method (provided drpcMethodHandlers are managed or accesible by it).

There you can also do other general checks, like the existence of a record with the given threadId (as stated in another comment) and verify that every request has an id in the case of a batch request.